File: multi_agent/projects/im-backend/src/main/java/com/im/server/announcement/Announcement.java
package com.im.server.announcement;

import java.time.LocalDateTime;
import java.util.*;

/**
 * 群公告实体
 * 支持Markdown富文本、置顶、编辑、历史记录
 */
public class Announcement {
    
    public enum AnnouncementType {
        NORMAL,      // 普通公告
        IMPORTANT,   // 重要公告
        PINNED       // 置顶公告
    }
    
    private String announcementId;
    private String groupId;
    private String authorId;
    private String authorName;
    private String title;
    private String content;          // Markdown内容
    private AnnouncementType type;
    private boolean pinned;           // 是否置顶
    private boolean edited;           // 是否被编辑过
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;
    private LocalDateTime pinnedAt;    // 置顶时间
    private int viewCount;            // 查看次数
    private int confirmCount;         // 确认人数
    private Set<String> confirmedUserIds; // 已确认用户
    private boolean deleted;
    private LocalDateTime deletedAt;

    public Announcement() {
        this.announcementId = UUID.randomUUID().toString();
        this.type = AnnouncementType.NORMAL;
        this.pinned = false;
        this.edited = false;
        this.createdAt = LocalDateTime.now();
        this.viewCount = 0;
        this.confirmCount = 0;
        this.confirmedUserIds = new HashSet<>();
        this.deleted = false;
    }

    // Builder模式
    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        private final Announcement announcement = new Announcement();

        public Builder groupId(String groupId) {
            announcement.groupId = groupId;
            return this;
        }

        public Builder authorId(String authorId) {
            announcement.authorId = authorId;
            return this;
        }

        public Builder authorName(String authorName) {
            announcement.authorName = authorName;
            return this;
        }

        public Builder title(String title) {
            announcement.title = title;
            return this;
        }

        public Builder content(String content) {
            announcement.content = content;
            return this;
        }

        public Builder type(AnnouncementType type) {
            announcement.type = type;
            return this;
        }

        public Builder pinned(boolean pinned) {
            announcement.pinned = pinned;
            return this;
        }

        public Announcement build() {
            return announcement;
        }
    }

    // Getters and Setters
    public String getAnnouncementId() {
        return announcementId;
    }

    public void setAnnouncementId(String announcementId) {
        this.announcementId = announcementId;
    }

    public String getGroupId() {
        return groupId;
    }

    public void setGroupId(String groupId) {
        this.groupId = groupId;
    }

    public String getAuthorId() {
        return authorId;
    }

    public void setAuthorId(String authorId) {
        this.authorId = authorId;
    }

    public String getAuthorName() {
        return authorName;
    }

    public void setAuthorName(String authorName) {
        this.authorName = authorName;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
    }

    public AnnouncementType getType() {
        return type;
    }

    public void setType(AnnouncementType type) {
        this.type = type;
    }

    public boolean isPinned() {
        return pinned;
    }

    public void setPinned(boolean pinned) {
        this.pinned = pinned;
        if (pinned) {
            this.pinnedAt = LocalDateTime.now();
        }
    }

    public boolean isEdited() {
        return edited;
    }

    public void setEdited(boolean edited) {
        this.edited = edited;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }

    public LocalDateTime getPinnedAt() {
        return pinnedAt;
    }

    public void setPinnedAt(LocalDateTime pinnedAt) {
        this.pinnedAt = pinnedAt;
    }

    public int getViewCount() {
        return viewCount;
    }

    public void setViewCount(int viewCount) {
        this.viewCount = viewCount;
    }

    public int getConfirmCount() {
        return confirmCount;
    }

    public void setConfirmCount(int confirmCount) {
        this.confirmCount = confirmCount;
    }

    public Set<String> getConfirmedUserIds() {
        return confirmedUserIds;
    }

    public void setConfirmedUserIds(Set<String> confirmedUserIds) {
        this.confirmedUserIds = confirmedUserIds;
    }

    public boolean isDeleted() {
        return deleted;
    }

    public void setDeleted(boolean deleted) {
        this.deleted = deleted;
        if (deleted) {
            this.deletedAt = LocalDateTime.now();
        }
    }

    public LocalDateTime getDeletedAt() {
        return deletedAt;
    }

    public void setDeletedAt(LocalDateTime deletedAt) {
        this.deletedAt = deletedAt;
    }

    // 业务方法
    public void incrementViewCount() {
        this.viewCount++;
    }

    public boolean confirm(String userId) {
        if (confirmedUserIds.add(userId)) {
            this.confirmCount = confirmedUserIds.size();
            return true;
        }
        return false;
    }

    public boolean unconfirm(String userId) {
        if (confirmedUserIds.remove(userId)) {
            this.confirmCount = confirmedUserIds.size();
            return true;
        }
        return false;
    }

    public boolean hasConfirmed(String userId) {
        return confirmedUserIds.contains(userId);
    }

    public void updateContent(String title, String content) {
        this.title = title;
        this.content = content;
        this.updatedAt = LocalDateTime.now();
        this.edited = true;
    }

    public void pin() {
        this.pinned = true;
        this.pinnedAt = LocalDateTime.now();
        this.type = AnnouncementType.PINNED;
    }

    public void unpin() {
        this.pinned = false;
        this.pinnedAt = null;
        this.type = AnnouncementType.NORMAL;
    }

    public void softDelete() {
        this.deleted = true;
        this.deletedAt = LocalDateTime.now();
    }

    // 转为Map用于JSON序列化
    public Map<String, Object> toMap() {
        Map<String, Object> map = new HashMap<>();
        map.put("announcementId", announcementId);
        map.put("groupId", groupId);
        map.put("authorId", authorId);
        map.put("authorName", authorName);
        map.put("title", title);
        map.put("content", content);
        map.put("type", type.name());
        map.put("pinned", pinned);
        map.put("edited", edited);
        map.put("createdAt", createdAt != null ? createdAt.toString() : null);
        map.put("updatedAt", updatedAt != null ? updatedAt.toString() : null);
        map.put("pinnedAt", pinnedAt != null ? pinnedAt.toString() : null);
        map.put("viewCount", viewCount);
        map.put("confirmCount", confirmCount);
        map.put("deleted", deleted);
        return map;
    }
}

File: multi_agent/projects/im-mobile/lib/models/announcement.dart
/**
 * 群公告模型 - 移动端 Flutter
 */

enum AnnouncementType { normal, important, pinned }

class Announcement {
    final String announcementId;
    final String groupId;
    final String authorId;
    final String authorName;
    final String title;
    final String content;
    final AnnouncementType type;
    final bool pinned;
    final bool edited;
    final DateTime createdAt;
    final DateTime? updatedAt;
    final DateTime? pinnedAt;
    final int viewCount;
    final int confirmCount;
    final Set<String> confirmedUserIds;
    final bool deleted;

    Announcement({
        required this.announcementId,
        required this.groupId,
        required this.authorId,
        required this.authorName,
        required this.title,
        required this.content,
        required this.type,
        this.pinned = false,
        this.edited = false,
        required this.createdAt,
        this.updatedAt,
        this.pinnedAt,
        this.viewCount = 0,
        this.confirmCount = 0,
        this.confirmedUserIds = const {},
        this.deleted = false,
    });

    factory Announcement.fromJson(Map<String, dynamic> json) {
        return Announcement(
            announcementId: json['announcementId'] ?? '',
            groupId: json['groupId'] ?? '',
            authorId: json['authorId'] ?? '',
            authorName: json['authorName'] ?? '',
            title: json['title'] ?? '',
            content: json['content'] ?? '',
            type: _parseType(json['type']),
            pinned: json['pinned'] ?? false,
            edited: json['edited'] ?? false,
            createdAt: _parseDateTime(json['createdAt']),
            updatedAt: _parseDateTimeNullable(json['updatedAt']),
            pinnedAt: _parseDateTimeNullable(json['pinnedAt']),
            viewCount: json['viewCount'] ?? 0,
            confirmCount: json['confirmCount'] ?? 0,
            confirmedUserIds: (json['confirmedUserIds'] as List<dynamic>?)
                    ?.map((e) => e.toString())
                    .toSet() ??
                {},
            deleted: json['deleted'] ?? false,
        );
    }

    Map<String, dynamic> toJson() {
        return {
            'announcementId': announcementId,
            'groupId': groupId,
            'authorId': authorId,
            'authorName': authorName,
            'title': title,
            'content': content,
            'type': type.name.toUpperCase(),
            'pinned': pinned,
            'edited': edited,
            'createdAt': createdAt.toIso8601String(),
            'updatedAt': updatedAt?.toIso8601String(),
            'pinnedAt': pinnedAt?.to8601String(),
            'viewCount': viewCount,
            'confirmCount': confirmCount,
            'confirmedUserIds': confirmedUserIds.toList(),
            'deleted': deleted,
        };
    }

    Announcement copyWith({
        String? announcementId,
        String? groupId,
        String? authorId,
        String? authorName,
        String? title,
        String? content,
        AnnouncementType? type,
        bool? pinned,
        bool? edited,
        DateTime? createdAt,
        DateTime? updatedAt,
        DateTime? pinnedAt,
        int? viewCount,
        int? confirmCount,
        Set<String>? confirmedUserIds,
        bool? deleted,
    }) {
        return Announcement(
            announcementId: announcementId ?? this.announcementId,
            groupId: groupId ?? this.groupId,
            authorId: authorId ?? this.authorId,
            authorName: authorName ?? this.authorName,
            title: title ?? this.title,
            content: content ?? this.content,
            type: type ?? this.type,
            pinned: pinned ?? this.pinned,
            edited: edited ?? this.edited,
            createdAt: createdAt ?? this.createdAt,
            updatedAt: updatedAt ?? this.updatedAt,
            pinnedAt: pinnedAt ?? this.pinnedAt,
            viewCount: viewCount ?? this.viewCount,
            confirmCount: confirmCount ?? this.confirmCount,
            confirmedUserIds: confirmedUserIds ?? this.confirmedUserIds,
            deleted: deleted ?? this.deleted,
        );
    }

    bool hasConfirmed(String userId) => confirmedUserIds.contains(userId);

    String get typeLabel {
        switch (type) {
            case AnnouncementType.important:
                return '重要';
            case AnnouncementType.pinned:
                return '置顶';
            default:
                return '公告';
        }
    }
}

AnnouncementType _parseType(String? type) {
    switch (type?.toUpperCase()) {
        case 'IMPORTANT':
            return AnnouncementType.important;
        case 'PINNED':
            return AnnouncementType.pinned;
        default:
            return AnnouncementType.normal;
    }
}

DateTime _parseDateTime(String? s) {
    if (s == null) return DateTime.now();
    return DateTime.tryParse(s) ?? DateTime.now();
}

DateTime? _parseDateTimeNullable(String? s) {
    if (s == null) return null;
    return DateTime.tryParse(s);
}

class CreateAnnouncementRequest {
    final String groupId;
    final String authorId;
    final String authorName;
    final String title;
    final String content;
    final String type;

    CreateAnnouncementRequest({
        required this.groupId,
        required this.authorId,
        required this.authorName,
        required this.title,
        required this.content,
        this.type = 'NORMAL',
    });

    Map<String, dynamic> toJson() {
        return {
            'groupId': groupId,
            'authorId': authorId,
            'authorName': authorName,
            'title': title,
            'content': content,
            'type': type,
        };
    }
}

class PagedAnnouncements {
    final List<Announcement> announcements;
    final int total;
    final int page;
    final int pageSize;
    final int totalPages;

    PagedAnnouncements({
        required this.announcements,
        required this.total,
        required this.page,
        required this.pageSize,
        required this.totalPages,
    });

    factory PagedAnnouncements.fromJson(Map<String, dynamic> json) {
        return PagedAnnouncements(
            announcements: (json['announcements'] as List<dynamic>?)
                    ?.map((e) => Announcement.fromJson(e as Map<String, dynamic>))
                    .toList() ??
                [],
            total: json['total'] ?? 0,
            page: json['page'] ?? 1,
            pageSize: json['pageSize'] ?? 20,
            totalPages: json['totalPages'] ?? 1,
        );
    }
}

class AnnouncementStatistics {
    final int total;
    final int pinned;
    final int important;
    final int totalViews;

    AnnouncementStatistics({
        required this.total,
        required this.pinned,
        required this.important,
        required this.totalViews,
    });

    factory AnnouncementStatistics.fromJson(Map<String, dynamic> json) {
        return AnnouncementStatistics(
            total: json['total'] ?? 0,
            pinned: json['pinned'] ?? 0,
            important: json['important'] ?? 0,
            totalViews: json['totalViews'] ?? 0,
        );
    }
}
